File: js/src/0.0-Basics/0.3-build-up-logical-thinking/0.3.27-pattern.js
function log(val) {
    process.stdout.write(String(val));
}

function pattern_27() {
    let row = 10;
    let column = 10;

    let end = (row + column) / 4;

    for (let y = 0; y <= row; y++) {
        let element = column / 2;

        for (let x = 0; x <= column; x++) {
            if (y <= row / 2) {
                if (x <= column / 2) {
                    log(` ${element}`);

                    if (end - y != element) {
                        element = element - 1;
                    }
                } else if (x >= column / 2) {
                    if (x - end > element) {
                        element = element + 1;
                    }
                    log(` ${element}`);
                }
            } else if (y > row / 2) {
                let key = column - y;
                if (x <= column / 2) {
                    log(` ${element}`);

                    if (end - key != element) {
                        element = element - 1;
                    }
                } else if (x >= column / 2) {
                    if (x - end > element) {
                        element = element + 1;
                    }
                    log(` ${element}`);
                }
            }
        }
        console.log();
    }

    return 0;
}

console.clear();
pattern_27();

File: python/00-Basics/00.03-build-up-logical-thinking/0.3.27-pattern.py
from utils.clear import clear

def pattern_27():
    row = 10
    column = 10

    for i in range(row , -1, -1):
        for j in range(column, -1, -1):
            # left half
            if j >= column / 2:
                newJ = round(j - column / 2)
                # top left
                if i  >= row / 2:
                    newI = round(i - row / 2)
                    if newJ >= newI:
                        print(newJ, " ", end="")
                    else:
                        print(newI, " ", end="")
                # bottom left
                if i  < row / 2:
                    newI = round(row / 2 - i)
                    if newJ >= newI:
                        print(newJ, " ", end="")
                    else:
                        print(newI, " ", end="")
            elif j < column /2:
                newJ =  round(column / 2  - j)
                # top left
                if i  >= row / 2:
                    newI = round(i - row / 2)
                    if newJ >= newI:
                        print(newJ, " ", end="")
                    else:
                        print(newI, " ", end="")
                # bottom left
                if i  < row / 2:
                    newI = round(row / 2 - i)
                    if newJ >= newI:
                        print(newJ, " ", end="")
                    else:
                        print(newI, " ", end="")

        if i >= row / 2:
            newI = round(i - row / 2)
            print(" ", newI)
        elif i < row / 2:
            newI = round(row / 2 - i)
            print(" ", newI)



clear()
pattern_27()
